document "advanced" tool test debugging (#1108)
galaxyproject/planemo: docs/writing_how_do_i.rst

.. _Docker Hub: https://hub.docker.com/
.. _Planemo: http://planemo.readthedocs.org/
.. _Picard: http://broadinstitute.github.io/picard/

----------------------------------------------------------------------------
\.\.\. test tools against a package or container in a bioconda pull request?
----------------------------------------------------------------------------

> First, obtain the artifacts of the PR by sending the following message

Reviewer: ```suggestion
First, obtain the artifacts of the PR by adding this comment:
```

``@BiocondaBot please fetch artifacts``. In the reply one finds the links to the
built package and docker image. 

In order to test the tool with the package add the following to the planemo call::

     $ planemo test ... --conda_channels LINK_TO_PACKAGE,conda-forge,bioconda,defaults ...

For containerized testing the docker image needs to be loaded::

     $ curl  -L "LINK_TO_DOCKER_IMAGE.tar.gz" | gzip -dc | docker load

A planemo test will then simply use this image::

     $ planemo test ... --biocontainers --no_conda_auto_init ...

--------------------------------------
\.\.\. interactively debug tool tests?
--------------------------------------

It can be desirable to interactively debug a tool test. In order to do so, start ``planemo test``

> with the option ``--no_cleanup``. Inspect the output: After starting galaxy tests start. At the

Reviewer: ```suggestion
with the option ``--no_cleanup``. Inspect the output: After Galaxy starts up, the tests commence. At the
```

> start of each test one finds a message: ``( TOOL_ID ) > Test-X``. After a some upload jobs the

Reviewer: ```suggestion
start of each test one finds a message: ``( <TOOL_ID> ) > Test-N``. After some upload jobs, the
```

> actual tool job is started (its the last before the next test is executed). There you will find

Reviewer: ```suggestion
actual tool job is started (it is the last before the next test is executed). There you will find
```

a message like ``Built script [/tmp/tmp1zixgse3/job_working_directory/000/3/tool_script.sh]``

In this case ``/tmp/tmp1zixgse3/job_working_directory/000/3/`` is the job dir. It contains some
files and directories of interest: 

> - ``tool_script.sh`` the bash script generated from the tool's ``command`` and ``version_command``

Reviewer: ```suggestion
- ``tool_script.sh``: the bash script generated from the tool's ``command`` and ``version_command``
```

  tags plus some boiler plate code

> - ``galaxy_3.sh`` (note that the number may be different) a schell script setting up the environment

Reviewer: ```suggestion
- ``galaxy_3.sh`` (note that the number may be different): a shell script setting up the environment
```

  (e.g. paths and environment variables), starting the ``tool_script.sh``, and postprocessing
  (e.g. error handling and setting metadata)

> - ``working`` the jobs working dir
> - ``outputs`` a dir containing the jobs stderr and stdout

Reviewer: ```suggestion
- ``working``: the job working directory
- ``outputs``: a directory containing the job stderr and stdout
```

For a tool test that uses a conda environment to resolve the requirements one can simply change
into ``working`` and execute ``../tool_script.sh`` (works as long as no special environment variables
are used; in this case ``../galaxy_3.sh`` needs to be executed after cleaning the job dir). 
By editing the tool script one may understand/fix problems in the ``command`` block faster than by
rerunning ``planemo test`` over and over again.

Alternatively one can change into the ``working`` dir and load the conda environment
(the code to do so can be found in ``tool_script.sh``: ``. PATH_TO_CONDA_ENV activate``). 

> Afterwards one can execute individual commands, e.g. those found in ``tool_script.sh`` or varaints.

Reviewer: ```suggestion
Afterwards one can execute individual commands, e.g. those found in ``tool_script.sh`` or variants.
```

> For a tool test that uses a docker to to resolve the requirements one needs to execute 

Reviewer: ```suggestion
For a tool test that uses Docker to to resolve the requirements one needs to execute 
```
